File: go-bindings/schemes.cpp
// Copyright (c) 2021 The Dash Core developers

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <vector>
#include "chiabls/bls.hpp"
#include "privatekey.h"
#include "elements.h"
#include "schemes.h"
#include "blschia.h"
#include "error.h"

// helper functions
std::vector<bls::G1Element> toVectorG1Elements(void** elems, const size_t len) {
    std::vector<bls::G1Element> vec;
    for (int i = 0 ; i < len; ++i) {
        const bls::G1Element* el = (bls::G1Element*)elems[i];
        vec.push_back(*el);
    }
    return vec;
}

std::vector<bls::G2Element> toVectorG2Elements(void** elems, const size_t len) {
    std::vector<bls::G2Element> vec;
    for (int i = 0 ; i < len; ++i) {
        const bls::G2Element* el = (bls::G2Element*)elems[i];
        vec.push_back(*el);
    }
    return vec;
}

std::vector<bls::Bytes> toVectorBytes(void** elems, const size_t len, const std::vector<size_t> vecElemsLens) {
    std::vector<bls::Bytes> vec;
    for (int i = 0 ; i < len; ++i) {
        uint8_t* elPtr = (uint8_t*)elems[i];
        vec.push_back(bls::Bytes(elPtr, vecElemsLens[i]));
    }
    return vec;
}

// Implementation of bindings for CoreMPL class

CPrivateKey CCoreMPLKeyGen(const CCoreMPL scheme, const void* seed, const size_t seedLen, bool* didErr) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    bls::PrivateKey* sk = nullptr;
    try {
        sk = new bls::PrivateKey(schemePtr->KeyGen(bls::Bytes((uint8_t*)seed, seedLen)));
    } catch (const std::exception& ex) {
        gErrMsg = ex.what();
        *didErr = true;
        return nullptr;
    }
    *didErr = false;
    return sk;
}

CG1Element CCoreMPSkToG1(const CCoreMPL scheme, const CPrivateKey sk) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    const bls::PrivateKey* skPtr = (bls::PrivateKey*)sk;
    return new bls::G1Element(
        schemePtr->SkToG1(*skPtr)
    );
}

CG2Element CCoreMPLSign(CCoreMPL scheme, const CPrivateKey sk, const void* msg, const size_t msgLen) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    const bls::PrivateKey* skPtr = (bls::PrivateKey*)sk;
    return new bls::G2Element(
        schemePtr->Sign(*skPtr, bls::Bytes((uint8_t*)msg, msgLen))
    );
}

bool CCoreMPLVerify(const CCoreMPL scheme,
                    const CG1Element pk,
                    const void* msg,
                    const size_t msgLen,
                    const CG2Element sig) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    const bls::G1Element* pkPtr = (bls::G1Element*)pk;
    const bls::G2Element* sigPtr = (bls::G2Element*)sig;
    return schemePtr->Verify(*pkPtr, bls::Bytes((uint8_t*)msg, msgLen), *sigPtr);
}

CG1Element CCoreMPLAggregatePubKeys(const CCoreMPL scheme, void** pks, const size_t pksLen) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    return new bls::G1Element(schemePtr->Aggregate(toVectorG1Elements(pks, pksLen)));
}

CG2Element CCoreMPLAggregateSigs(const CCoreMPL scheme, void** sigs, const size_t sigsLen) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    return new bls::G2Element(
        schemePtr->Aggregate(toVectorG2Elements(sigs, sigsLen))
    );
}

CPrivateKey CCoreMPLDeriveChildSk(const CCoreMPL scheme, const CPrivateKey sk, const uint32_t index) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    const bls::PrivateKey* skPtr = (bls::PrivateKey*)sk;
    return new bls::PrivateKey(schemePtr->DeriveChildSk(*skPtr, index));
}

CPrivateKey CCoreMPLDeriveChildSkUnhardened(CCoreMPL scheme, CPrivateKey sk, uint32_t index) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    bls::PrivateKey* skPtr = (bls::PrivateKey*)sk;
    return new bls::PrivateKey(schemePtr->DeriveChildSkUnhardened(*skPtr, index));
}

CG1Element CCoreMPLDeriveChildPkUnhardened(CCoreMPL scheme, CG1Element el, uint32_t index) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    bls::G1Element* elPtr = (bls::G1Element*)el;
    return new bls::G1Element(schemePtr->DeriveChildPkUnhardened(*elPtr, index));
}

bool CCoreMPLAggregateVerify(const CCoreMPL scheme,
                             void** pks,
                             const size_t pksLen,
                             void** msgs,
                             const void* msgsLens,
                             const size_t msgsLen,
                             const CG2Element sig) {
    bls::CoreMPL* schemePtr = (bls::CoreMPL*)scheme;
    const size_t* msgLensPtr = (size_t*)msgsLens;
    const bls::G2Element* sigPtr = (bls::G2Element*)sig;
    const std::vector<bls::G1Element> vecPubKeys = toVectorG1Elements(pks, pksLen);
    const std::vector<size_t> vecMsgsLens = std::vector<size_t>(msgLensPtr, msgLensPtr + msgsLen);
    const std::vector<bls::Bytes> vecMsgs = toVectorBytes(msgs, msgsLen, vecMsgsLens);
    return schemePtr->AggregateVerify(vecPubKeys, vecMsgs, *sigPtr);
}

// BasicSchemeMPL
CBasicSchemeMPL NewCBasicSchemeMPL() {
    return new bls::BasicSchemeMPL();
}

bool CBasicSchemeMPLAggregateVerify(CBasicSchemeMPL scheme,
                                    void** pks,
                                    const size_t pksLen,
                                    void** msgs,
                                    const void* msgsLens,
                                    const size_t msgsLen,
                                    const CG2Element sig) {
    bls::BasicSchemeMPL* schemePtr = (bls::BasicSchemeMPL*)scheme;
    const size_t* msgLensPtr = (size_t*)msgsLens;
    const bls::G2Element* sigPtr = (bls::G2Element*)sig;
    const std::vector<bls::G1Element> vecPubKeys = toVectorG1Elements(pks, pksLen);
    const std::vector<size_t> vecMsgsLens = std::vector<size_t>(msgLensPtr, msgLensPtr + msgsLen);
    const std::vector<bls::Bytes> vecMsgs = toVectorBytes(msgs, msgsLen, vecMsgsLens);
    return schemePtr->AggregateVerify(vecPubKeys, vecMsgs, *sigPtr);
}

void CBasicSchemeMPLFree(CBasicSchemeMPL scheme) {
    bls::BasicSchemeMPL* schemePtr = (bls::BasicSchemeMPL*)scheme;
    delete schemePtr;
}

// AugSchemeMPL
CAugSchemeMPL NewCAugSchemeMPL() {
    return new bls::AugSchemeMPL();
}

CG2Element CAugSchemeMPLSign(const CAugSchemeMPL scheme, const CPrivateKey sk, const void* msg, const size_t msgLen) {
    bls::AugSchemeMPL* schemePtr = (bls::AugSchemeMPL*)scheme;
    const bls::PrivateKey* skPtr = (bls::PrivateKey*)sk;
    return new bls::G2Element(
        schemePtr->Sign(*skPtr, bls::Bytes((uint8_t*)msg, msgLen))
    );
}

CG2Element CAugSchemeMPLSignPrepend(const CAugSchemeMPL scheme,
                                    const CPrivateKey sk,
                                    const void* msg,
                                    const size_t msgLen,
                                    const CG1Element prepPk) {
    bls::AugSchemeMPL* schemePtr = (bls::AugSchemeMPL*)scheme;
    const bls::PrivateKey* skPtr = (bls::PrivateKey*)sk;
    const bls::G1Element* prepPkPtr = (bls::G1Element*)prepPk;
    return new bls::G2Element(schemePtr->Sign(*skPtr, bls::Bytes((uint8_t*)msg, msgLen), *prepPkPtr));
}

bool CAugSchemeMPLVerify(const CAugSchemeMPL scheme,
                         const CG1Element pk,
                         const void* msg,
                         const size_t msgLen,
                         const CG2Element sig) {
    bls::AugSchemeMPL* schemePtr = (bls::AugSchemeMPL*)scheme;
    const bls::G1Element* pkPtr = (bls::G1Element*)pk;
    const uint8_t* msgPtr = (uint8_t*)msg;
    const bls::G2Element* sigPtr = (bls::G2Element*)sig;
    return schemePtr->Verify(*pkPtr, bls::Bytes(msgPtr, msgLen), *sigPtr);
}

bool CAugSchemeMPLAggregateVerify(const CAugSchemeMPL scheme,
                                  void** pks,
                                  const size_t pksLen,
                                  void** msgs,
                                  const void* msgsLens,
                                  const size_t msgsLen,
                                  const CG2Element sig) {
    bls::AugSchemeMPL* schemePtr = (bls::AugSchemeMPL*)scheme;
    const size_t* msgLensPtr = (size_t*)msgsLens;
    const bls::G2Element* sigPtr = (bls::G2Element*)sig;
    const std::vector<bls::G1Element> vecPubKeys = toVectorG1Elements(pks, pksLen);
    const std::vector<size_t> vecMsgsLens = std::vector<size_t>(msgLensPtr, msgLensPtr + msgsLen);
    const std::vector<bls::Bytes> vecMsgs = toVectorBytes(msgs, msgsLen, vecMsgsLens);
    return schemePtr->AggregateVerify(vecPubKeys, vecMsgs, *sigPtr);
}

void CAugSchemeMPLFree(CAugSchemeMPL scheme) {
    bls::AugSchemeMPL* schemePtr = (bls::AugSchemeMPL*)scheme;
    delete schemePtr;
}

// PopSchemeMPL
CPopSchemeMPL NewCPopSchemeMPL() {
    return new bls::PopSchemeMPL();
}

CG2Element CPopSchemeMPLPopProve(const CPopSchemeMPL scheme, const CPrivateKey sk) {
    bls::PopSchemeMPL* schemePtr = (bls::PopSchemeMPL*)scheme;
    const bls::PrivateKey* skPtr = (bls::PrivateKey*)sk;
    return new bls::G2Element(schemePtr->PopProve(*skPtr));
}

bool CPopSchemeMPLPopVerify(const CPopSchemeMPL scheme, const CG1Element pk, const CG2Element sig) {
    bls::PopSchemeMPL* schemePtr = (bls::PopSchemeMPL*)scheme;
    const bls::G1Element* pkPtr = (bls::G1Element*)pk;
    const bls::G2Element* sigPtr = (bls::G2Element*)sig;
    return schemePtr->PopVerify(*pkPtr, *sigPtr);
}

bool CPopSchemeMPLFastAggregateVerify(const CPopSchemeMPL scheme,
                                      void** pks,
                                      const size_t pksLen,
                                      const void* msg,
                                      const size_t msgLen,
                                      const CG2Element sig) {
    bls::PopSchemeMPL* schemePtr = (bls::PopSchemeMPL*)scheme;
    const bls::G2Element* sigPtr = (bls::G2Element*)sig;
    const std::vector<bls::G1Element> vecPubKeys = toVectorG1Elements(pks, pksLen);
    return schemePtr->FastAggregateVerify(vecPubKeys, bls::Bytes((uint8_t*)msg, msgLen), *sigPtr);
}

void CPopSchemeMPLFree(CPopSchemeMPL scheme) {
    bls::PopSchemeMPL* schemePtr = (bls::PopSchemeMPL*)scheme;
    delete schemePtr;
}

File: flutter-bindings/bls_signatures_ffi/lib/src/schemes.dart
import 'dart:ffi';
import 'dart:typed_data';

import 'package:bls_signatures_ffi/bls_signatures_ffi.dart';
import 'package:bls_signatures_ffi/src/blschia.dart';
import 'package:bls_signatures_ffi/src/blschia.h.dart';
import 'package:ffi/ffi.dart';
import 'package:flutter/foundation.dart';

/// An instance of [CoreMPL].
abstract class CoreMPL {
  /// @nodoc
  @protected
  const CoreMPL(CCoreMPL coreMPL) : _coreMPL = coreMPL;

  final CCoreMPL _coreMPL;

  /// Create a [PrivateKey] instance from [seed].
  PrivateKey keyGen(Uint8List seed) {
    final seedBuf = malloc.allocate<Uint8>(seed.length)
      ..asTypedList(seed.length).setAll(0, seed);

    final didErr = malloc.allocate<Bool>(1);
    final pk = PrivateKey(
      bindings.CCoreMPLKeyGen(
        _coreMPL,
        seedBuf.cast(),
        seed.length,
        didErr.cast(),
      ),
    );
    malloc.free(seedBuf);
    if (didErr.value) {
      malloc.free(didErr);
      throw BLSException.errFromC();
    }
    malloc.free(didErr);
    return pk;
  }

  /// Convert [PrivateKey] into [G1Element].
  G1Element skToG1(PrivateKey privateKey) {
    // ignore: invalid_use_of_protected_member
    return G1Element(bindings.CCoreMPSkToG1(_coreMPL, privateKey.ptr));
  }

  /// Sign a [msg] using a [privateKey].
  G2Element sign(PrivateKey privateKey, Uint8List msg) {
    final msgBuf = malloc.allocate<Uint8>(msg.length)
      ..asTypedList(msg.length).setAll(0, msg);
    final sig = G2Element(
      bindings.CCoreMPLSign(
        _coreMPL,
        // ignore: invalid_use_of_protected_member
        privateKey.ptr,
        msgBuf.cast(),
        msg.length,
      ),
    );
    malloc.free(msgBuf);
    return sig;
  }

  /// Verify a [sig] for a [msg] using [g1Element] as public key.
  bool verify(G1Element g1Element, Uint8List msg, G2Element sig) {
    final msgBuf = malloc.allocate<Uint8>(msg.length)
      ..asTypedList(msg.length).setAll(0, msg);
    final isVerified = bindings.CCoreMPLVerify(
      _coreMPL,
      // ignore: invalid_use_of_protected_member
      g1Element.ptr,
      msgBuf.cast(),
      msg.length,
      // ignore: invalid_use_of_protected_member
      sig.ptr,
    );
    malloc.free(msgBuf);
    return isVerified;
  }

  /// Aggregate list of [G1Element].
  G1Element aggregatePubKeys(List<G1Element> pks) {
    final arrBuf = bindings.AllocPtrArray(pks.length);
    pks.asMap().forEach((index, pk) {
      // ignore: invalid_use_of_protected_member
      bindings.SetPtrArray(arrBuf.cast(), pk.ptr, index);
    });
    final aggSig = G1Element(
      bindings.CCoreMPLAggregatePubKeys(
        _coreMPL,
        arrBuf.cast(),
        pks.length,
      ),
    );
    malloc.free(arrBuf);
    return aggSig;
  }

  /// Aggregate list of [G2Element].
  G2Element aggregateSigs(List<G2Element> sigs) {
    final arrBuf = malloc.allocate<Uint32>(sigs.length);
    sigs.asMap().forEach((index, sig) {
      // ignore: invalid_use_of_protected_member
      bindings.SetPtrArray(arrBuf.cast(), sig.ptr, index);
    });
    final aggSig = G2Element(
      bindings.CCoreMPLAggregateSigs(
        _coreMPL,
        arrBuf.cast(),
        sigs.length,
      ),
    );
    malloc.free(arrBuf);
    return aggSig;
  }

  /// Get a [privateKey]'s child.
  PrivateKey deriveChildSk(PrivateKey privateKey, int index) {
    return PrivateKey(
      // ignore: invalid_use_of_protected_member
      bindings.CCoreMPLDeriveChildSk(_coreMPL, privateKey.ptr, index),
    );
  }

  /// Get a [privateKey]'s child.
  PrivateKey deriveChildSkUnhardened(PrivateKey privateKey, int index) {
    return PrivateKey(
      // ignore: invalid_use_of_protected_member
      bindings.CCoreMPLDeriveChildSkUnhardened(_coreMPL, privateKey.ptr, index),
    );
  }

  /// Get a [g1]'s child.
  G1Element deriveChildPkUnhardened(G1Element g1, int index) {
    return G1Element(
      // ignore: invalid_use_of_protected_member
      bindings.CCoreMPLDeriveChildPkUnhardened(_coreMPL, g1.ptr, index),
    );
  }

  /// Verify the aggregated signature [sig] for a list of messages [msgs]
  /// with list of public keys [pks].
  bool aggregateVerify(
    List<G1Element> pks,
    List<Uint8List> msgs,
    G2Element sig,
  ) {
    final pkBuf = bindings.AllocPtrArray(pks.length);
    pks.asMap().forEach((index, pk) {
      // ignore: invalid_use_of_protected_member
      bindings.SetPtrArray(pkBuf.cast(), pk.ptr, index);
    });
    final msgBuf = bindings.AllocPtrArray(msgs.length);
    final msgLens = <int>[];
    final msgPtrBuf = <Pointer<Uint8>>[];
    msgs.asMap().forEach((index, msg) {
      final mBuf = malloc.allocate<Uint8>(msg.length)
        ..asTypedList(msg.length).setAll(0, msg);
      msgPtrBuf.add(mBuf);
      msgLens.add(msg.length);
      bindings.SetPtrArray(msgBuf.cast(), mBuf.cast(), index);
    });
    final msgLenBuf = malloc.allocate<Int64>(msgLens.length)
      ..asTypedList(msgLens.length).setAll(0, msgLens);
    final res = bindings.CCoreMPLAggregateVerify(
      _coreMPL,
      pkBuf.cast(),
      pks.length,
      msgBuf.cast(),
      msgLenBuf.cast(),
      msgs.length,
      // ignore: invalid_use_of_protected_member
      sig.ptr,
    );

    malloc.free(pkBuf);
    // ignore: cascade_invocations
    malloc.free(msgBuf);
    for (final buf in msgPtrBuf) {
      malloc.free(buf);
    }
    malloc.free(msgLenBuf);
    return res;
  }
}

/// An instance of [BasicSchemeMPL].
class BasicSchemeMPL extends CoreMPL {
  /// Create an instance of [BasicSchemeMPL].
  BasicSchemeMPL() : super(bindings.NewCBasicSchemeMPL());

  @override
  bool aggregateVerify(
    List<G1Element> pks,
    List<Uint8List> msgs,
    G2Element sig,
  ) {
    final pkBuf = bindings.AllocPtrArray(pks.length);
    pks.asMap().forEach((index, pk) {
      // ignore: invalid_use_of_protected_member
      bindings.SetPtrArray(pkBuf.cast(), pk.ptr, index);
    });
    final msgBuf = bindings.AllocPtrArray(msgs.length);
    final msgLens = <int>[];
    final msgPtrBuf = <Pointer<Uint8>>[];
    msgs.asMap().forEach((index, msg) {
      final mBuf = malloc.allocate<Uint8>(msg.length)
        ..asTypedList(msg.length).setAll(0, msg);
      msgPtrBuf.add(mBuf);
      msgLens.add(msg.length);
      bindings.SetPtrArray(msgBuf.cast(), mBuf.cast(), index);
    });
    final msgLenBuf = malloc.allocate<Int64>(msgLens.length)
      ..asTypedList(msgLens.length).setAll(0, msgLens);
    final res = bindings.CBasicSchemeMPLAggregateVerify(
      _coreMPL,
      pkBuf.cast(),
      pks.length,
      msgBuf.cast(),
      msgLenBuf.cast(),
      msgs.length,
      // ignore: invalid_use_of_protected_member
      sig.ptr,
    );

    malloc.free(pkBuf);
    // ignore: cascade_invocations
    malloc.free(msgBuf);
    for (final buf in msgPtrBuf) {
      malloc.free(buf);
    }
    malloc.free(msgLenBuf);
    return res;
  }

  /// Release a [BasicSchemeMPL] instance from memory.
  void free() {
    bindings.CBasicSchemeMPLFree(_coreMPL);
  }
}

/// An instance of [AugSchemeMPL].
class AugSchemeMPL extends CoreMPL {
  /// Create an instance of [AugSchemeMPL].
  AugSchemeMPL() : super(bindings.NewCAugSchemeMPL());

  @override
  G2Element sign(PrivateKey privateKey, Uint8List msg) {
    final msgBuf = malloc.allocate<Uint8>(msg.length)
      ..asTypedList(msg.length).setAll(0, msg);
    final sig = G2Element(
      bindings.CAugSchemeMPLSign(
        _coreMPL,
        // ignore: invalid_use_of_protected_member
        privateKey.ptr,
        msgBuf.cast(),
        msg.length,
      ),
    );
    malloc.free(msgBuf);
    return sig;
  }

  /// Prepending different message.
  G2Element signPrepend(
    PrivateKey privateKey,
    Uint8List msg,
    G1Element prepPK,
  ) {
    final msgBuf = malloc.allocate<Uint8>(msg.length)
      ..asTypedList(msg.length).setAll(0, msg);
    final sig = G2Element(
      bindings.CAugSchemeMPLSignPrepend(
        _coreMPL,
        // ignore: invalid_use_of_protected_member
        privateKey.ptr,
        msgBuf.cast(),
        msg.length,
        // ignore: invalid_use_of_protected_member
        prepPK.ptr,
      ),
    );
    malloc.free(msgBuf);
    return sig;
  }

  @override
  bool verify(G1Element g1Element, Uint8List msg, G2Element sig) {
    final msgBuf = malloc.allocate<Uint8>(msg.length)
      ..asTypedList(msg.length).setAll(0, msg);
    final isVerified = bindings.CAugSchemeMPLVerify(
      _coreMPL,
      // ignore: invalid_use_of_protected_member
      g1Element.ptr,
      msgBuf.cast(),
      msg.length,
      // ignore: invalid_use_of_protected_member
      sig.ptr,
    );
    malloc.free(msgBuf);
    return isVerified;
  }

  @override
  bool aggregateVerify(
    List<G1Element> pks,
    List<Uint8List> msgs,
    G2Element sig,
  ) {
    final pkBuf = bindings.AllocPtrArray(pks.length);
    pks.asMap().forEach((index, pk) {
      // ignore: invalid_use_of_protected_member
      bindings.SetPtrArray(pkBuf.cast(), pk.ptr, index);
    });
    final msgBuf = bindings.AllocPtrArray(msgs.length);
    final msgLens = <int>[];
    final msgPtrBuf = <Pointer<Uint8>>[];
    msgs.asMap().forEach((index, msg) {
      final mBuf = malloc.allocate<Uint8>(msg.length)
        ..asTypedList(msg.length).setAll(0, msg);
      msgPtrBuf.add(mBuf);
      msgLens.add(msg.length);
      bindings.SetPtrArray(msgBuf.cast(), mBuf.cast(), index);
    });
    final msgLenBuf = malloc.allocate<Int64>(msgLens.length)
      ..asTypedList(msgLens.length).setAll(0, msgLens);
    final res = bindings.CAugSchemeMPLAggregateVerify(
      _coreMPL,
      pkBuf.cast(),
      pks.length,
      msgBuf.cast(),
      msgLenBuf.cast(),
      msgs.length,
      // ignore: invalid_use_of_protected_member
      sig.ptr,
    );

    malloc.free(pkBuf);
    // ignore: cascade_invocations
    malloc.free(msgBuf);
    for (final buf in msgPtrBuf) {
      malloc.free(buf);
    }
    malloc.free(msgLenBuf);
    return res;
  }

  /// Release a [AugSchemeMPL] instance from memory.
  void free() {
    bindings.CAugSchemeMPLFree(_coreMPL);
  }
}

/// An instance of [PopSchemeMPL].
class PopSchemeMPL extends CoreMPL {
  /// Create an instance of [PopSchemeMPL].
  PopSchemeMPL() : super(bindings.NewCPopSchemeMPL());

  /// Prove using [privateKey].
  G2Element popProve(PrivateKey privateKey) {
    return G2Element(
      // ignore: invalid_use_of_protected_member
      bindings.CPopSchemeMPLPopProve(_coreMPL, privateKey.ptr),
    );
  }

  /// Verifies [sig] using proof of possesion.
  bool popVerify(G1Element g1Element, G2Element sig) {
    // ignore: invalid_use_of_protected_member
    return bindings.CPopSchemeMPLPopVerify(_coreMPL, g1Element.ptr, sig.ptr);
  }

  /// Fast verification.
  bool fastAggregateVerify(
    List<G1Element> pks,
    Uint8List msg,
    G2Element sig,
  ) {
    final pkBuf = bindings.AllocPtrArray(pks.length);
    pks.asMap().forEach((index, pk) {
      // ignore: invalid_use_of_protected_member
      bindings.SetPtrArray(pkBuf.cast(), pk.ptr, index);
    });
    final msgBuf = malloc.allocate<Uint8>(msg.length)
      ..asTypedList(msg.length).setAll(0, msg);
    final res = bindings.CPopSchemeMPLFastAggregateVerify(
      _coreMPL,
      pkBuf.cast(),
      pks.length,
      msgBuf.cast(),
      msg.length,
      // ignore: invalid_use_of_protected_member
      sig.ptr,
    );

    malloc.free(pkBuf);
    // ignore: cascade_invocations
    malloc.free(msgBuf);
    return res;
  }

  /// Release a [PopSchemeMPL] instance from memory.
  void free() {
    bindings.CPopSchemeMPLFree(_coreMPL);
  }
}
